### PR TITLE
CXP-544: Do not save product/product models when their associations were not updated

### PR DESCRIPTION
**Pull request by willy-ahva**
**Description (for Contributor and Core Developer)**

In #13178 and #13338 we found out that knowing if an entity with associations (in our case, products and product models) was actually updated was pretty hard: they contain a collection of associations, which themselves contain 3 collections (products, product models and groups). The comparison between former associations and new ones could be really tedious.

In order to ease that, we decided to radically change the `EntityWithAssociationsInterface` for it to be closer to functional use cases, and to help respect the law of Demeter:
- remove the `setAssociations()` method (functionally speaking, we never want to set all of the associations at once, we usually want to associate or deassociate an entity to another one)
- remove ` getAssociationForType()` and `getAssociationForTypeCode()` methods (in the same way, we don't want to have the whole association, we usually just want to know which products/product models/groups are associated to our entity) - Note: we actually kept the `getAssociationForTypeCode()` method in the Product and ProductModel implementations, but it is now protected and only used internally)

we also added the following methods:
- `hasAssociationForTypeCode(string $associationTypeCode): bool`
- `addAssociatedProduct(ProductInterface $product, string $associationTypeCode): void;`
- `removeAssociatedProduct(ProductInterface $product, string $associationTypeCode): void;`
- `getAssociatedProducts(string $associationTypeCode): ?Collection;`
- `addAssociatedProductModel(ProductModelInterface $productModel, string $associationTypeCode): void;`
- `removeAssociatedProductModel(ProductModelInterface $productModel, string $associationTypeCode): void;`
- `getAssociatedProductModels(string $associationTypeCode): ?Collection;`
- `addAssociatedGroup(GroupInterface $group, string $associationTypeCode): void;`
- `removeAssociatedGroup(GroupInterface $group, string $associationTypeCode): void;`
- `getAssociatedGroups(string $associationTypeCode): ?Collection;`

All these changes allowed us to easily detect if the association's owner was updated or not (in this case, if an entity was actually associated to it/deassociated from it)

As a bonus, I find the associations manipulation code (AssociationFieldSetter, AssociationFieldAdder, ...) more readable and more easily understandable

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | OK
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

**Comment by mmetayer** [comment]
TODO: revert Dockerfile.dev

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php [hunk 1]
```diff
@@ -571,12 +553,110 @@ public function __toString()
...
+            $association->removeProduct($product);
+            $this->dirty = true;
+        }
+    }
+
+    public function getAssociatedProducts(string $associationTypeCode): ?Collection
+    {
+        $association = $this->getAssociationForTypeCode($associationTypeCode);
+
+        return $association ? $association->getProducts() : null;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php:589]: return a copy of the collection

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php [hunk 2]
```diff
@@ -571,12 +553,110 @@ public function __toString()
...
+        $association = $this->getAssociationForTypeCode($associationTypeCode);
+        if ($association instanceof AssociationInterface && $association->getProductModels()->contains($productModel)) {
+            $association->removeProductModel($productModel);
+            $this->dirty = true;
+        }
+    }
+
+    public function getAssociatedProductModels(string $associationTypeCode): ?Collection
+    {
+        $association = $this->getAssociationForTypeCode($associationTypeCode);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php:618]: same

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php [hunk 3]
```diff
@@ -571,12 +553,110 @@ public function __toString()
...
+
+        if ($association instanceof AssociationInterface && $association->getGroups()->contains($group)) {
+            $association->removeGroup($group);
+            $this->dirty = true;
+        }
+    }
+
+    public function getAssociatedGroups(string $associationTypeCode): ?Collection
+    {
+        $association = $this->getAssociationForTypeCode($associationTypeCode);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/AbstractProduct.php:649]: same

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/ProductModel.php [hunk 4]
```diff
@@ -728,6 +676,156 @@ public function __toString()
...
+
+        if ($association instanceof AssociationInterface && $association->getGroups()->contains($group)) {
+            $association->removeGroup($group);
+            $this->dirty = true;
+        }
+    }
+
+    public function getAssociatedGroups(string $associationTypeCode): ?Collection
+    {
+        $association = $this->getAssociationForTypeCode($associationTypeCode);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/ProductModel.php:813]: return copy

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/ProductModel.php [hunk 5]
```diff
@@ -728,6 +676,156 @@ public function __toString()
...
+
+        if ($association instanceof AssociationInterface && $association->hasProduct($product)) {
+            $association->removeProduct($product);
+            $this->dirty = true;
+        }
+    }
+
+    public function getAssociatedProducts(string $associationTypeCode): ?Collection
+    {
+        $association = $this->getAssociationForTypeCode($associationTypeCode);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/ProductModel.php:749]: return copy

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Clearer/Field/AssociationFieldClearer.php [hunk 6]
```diff
@@ -35,17 +38,41 @@ public function clear($entity, string $property, array $options = []): void
...
+        }
+
+        foreach ($entity->getAssociations() as $association) {
+            $associationType = $association->getAssociationType();
+            $typeCode = $associationType->getCode();
+
+            foreach ($entity->getAssociatedProducts($typeCode) as $associatedProduct) {
+                $entity->removeAssociatedProduct($associatedProduct, $typeCode);
+                if ($associationType->isTwoWay()) {
+                    $this->removeInversedAssociation($associatedProduct, $typeCode, $entity);
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Clearer/Field/AssociationFieldClearer.php:52]: todo: use twowayassociationupdater

File: src/Akeneo/Pim/Enrichment/Component/Product/Updater/Adder/AssociationFieldAdder.php [hunk 7]
```diff
@@ -71,50 +64,57 @@ public function __construct(
      *     },
      * }
      */
-    public function addFieldData($product, $field, $data, array $options = [])
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Updater/Adder/AssociationFieldAdder.php:74]: Use TwoWayAssociationFieldUpdater

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Updater/TwoWayAssociationUpdater.php [hunk 8]
```diff
@@ -28,119 +28,77 @@ public function __construct(
...
+     * To fix this, we look for cloned objects by comparing the identifier.
+     *
      * {@inheritdoc}
      */
     public function createInversedAssociation(
-        AssociationInterface $association,
-        EntityWithAssociationsInterface $owner
+        EntityWithAssociationsInterface $owner,
+        string $associationTypeCode,
+        $associatedEntity
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Updater/TwoWayAssociationUpdater.php:41]: ?
```suggestion
        EntityWithAssociationsInterface $associatedEntity
```

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Updater/TwoWayAssociationUpdater.php [hunk 9]
```diff
@@ -28,119 +28,77 @@ public function __construct(
...
-
     /**
      * {@inheritdoc}
      */
     public function removeInversedAssociation(
-        AssociationInterface $association,
-        EntityWithAssociationsInterface $owner
+        EntityWithAssociationsInterface $owner,
+        string $associationTypeCode,
+        $associatedEntity
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Updater/TwoWayAssociationUpdater.php:80]:
```suggestion
        EntityWithAssociationsInterface $associatedEntity
```

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Updater/TwoWayAssociationUpdater.php [hunk 10]
```diff
@@ -28,119 +28,77 @@ public function __construct(
...
     ): void {
-        $associationType = $association->getAssociationType();
-        $associatedEntityToRemove = $association->getOwner();
-
-        /** @var AssociationInterface $inversedAssociation */
-        $inversedAssociation = $owner->getAssociationForType($associationType);
-        if (null === $inversedAssociation) {
-            return;
+        if (!$associatedEntity->hasAssociationForTypeCode($associationTypeCode)) {
+            $this->missingAssociationAdder->addMissingAssociations($associatedEntity);
```
Comment (StevenVAIDIE) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Updater/TwoWayAssociationUpdater.php:83]: Not really useful

File: tests/back/Pim/Enrichment/Specification/Component/Product/Model/ProductModelSpec.php [hunk 11]
```diff
@@ -717,6 +720,358 @@ function it_is_updated_when_adding_a_non_empty_association(
...
+        $xsellAssociation = new ProductAssociation();
+        $xsellType = new AssociationType();
+        $xsellType->setCode('x_sell');
+        $xsellAssociation->setAssociationType($xsellType);
+        $xsellAssociation->addProduct($plate);
+        $xsellAssociation->addProduct($spoon);
+
+        $this->addAssociation($xsellAssociation);
+
+        $this->getAssociatedProducts('x_sell')->shouldBeLike(new ArrayCollection([$plate, $spoon]));
```
Comment (StevenVAIDIE) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Model/ProductModelSpec.php:848]: Same for getAssociatedProductModels and getAssociatedGroups and in ProductSpec

```suggestion
        $this->getAssociatedProducts('x_sell')->shouldBeLike(new ArrayCollection([$plate, $spoon]));
        $this->getAssociatedProducts('another_association_type_code')->shouldBeLike(new ArrayCollection([$plate, $spoon])); 
```

**Review by StevenVAIDIE — COMMENTED** [review]
Really nice PR congrat ! :clap:

File: CHANGELOG.md [hunk 12]
```diff
@@ -234,7 +234,28 @@
...
+    - Add method `removeAssociatedProduct()`
+    - Add method `getAssociatedProducts()`
+    - Add method `addAssociatedProductModel()`
+    - Add method `removeAssociatedProductModel()`
+    - Add method `getAssociatedProductModels()`
+    - Add method `addAssociatedGroup()`
+    - Add method `removeAssociatedGroup()`
+    - Add method `getAssociatedGroups()`
+- Change constructor of `Akeneo\Pim\Enrichment\Component\Product\Updater\Adder\AssociationFieldAdder` argument `Akeneo\Pim\Structure\Component\Repository\AssociationTypeRepositoryInterface $associationTypeRepository` has been added
+- Change constructor of `Akeneo\Pim\Enrichment\Component\Product\Updater\Setter\AssociationFieldSetter` argument `Akeneo\Pim\Structure\Component\Repository\AssociationTypeRepositoryInterface $associationTypeRepository` has been added
```
Comment (mmetayer) [inline CHANGELOG.md:258]: update changelog for AssociationFieldClearer and AssociationFieldAdder

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Updater/TwoWayAssociationUpdater.php (hunk 8 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Updater/TwoWayAssociationUpdater.php:41]: you're right, we mixed up the signature of the arguments: $associatedEntity is a EntityWithAssociationsInterface, but $owner is actually a ProductInterface|ProductModelInterface. The union type is not available yet in PHP7.4, we will add it in PHP8

**Review by pierallard — APPROVED** [review]
You should write `wer` instead of `were` in the PR's title

**Review by nmarniesse — APPROVED** [review]

**Review by StevenVAIDIE — APPROVED** [review]

**Pull request merged**